lottery/calypso/mod.go

```diff
@@ -0,0 +1,238 @@
+package calypso
+
+import (
+	"crypto/rand"
+	"encoding/json"
+
+	"go.dedis.ch/dela/serde"
+	serdej "go.dedis.ch/dela/serde/json"
+
+	"go.dedis.ch/dela/dkg"
+	"go.dedis.ch/dela/ledger/arc"
+	"go.dedis.ch/dela/ledger/arc/darc"
+	"go.dedis.ch/dela/lottery"
+	"go.dedis.ch/dela/lottery/storage"
+	"go.dedis.ch/dela/lottery/storage/inmemory"
+	"go.dedis.ch/dela/mino"
+	"go.dedis.ch/kyber/v3"
+	"go.dedis.ch/kyber/v3/suites"
+	"golang.org/x/xerrors"
+)
+
+const (
+	keySize = 32
+
+	// ArcRuleUpdate defines the rule to update the arc. This rule must be set
+	// at the write creation to allow the arc to be latter updated.
+	ArcRuleUpdate = "calypso_update"
+	// ArcRuleRead defines the arc rule to read a value
+	ArcRuleRead = "calypso_read"
+)
+
+// Suite is the Kyber suite for Pedersen.
+var suite = suites.MustFind("Ed25519")
+
+// NewCalypso creates a new Calypso
+func NewCalypso(dkg dkg.DKG) *Calypso {
+	return &Calypso{
+		dkg:        dkg,
+		storage:    inmemory.NewInMemory(),
+		serializer: serdej.NewSerializer(),
+	}
+}
+
+// Calypso is a wrapper around DKG to provide a storage and authorization layer
+//
+// implements lottery.Secret
+type Calypso struct {
+	dkg        dkg.DKG
+	dkgActor   dkg.Actor
+	storage    storage.KeyValue
+	serializer serde.Serializer
+}
+
+// Setup implements lottery.Secret
+func (c *Calypso) Setup(players mino.Players, pubKeys []kyber.Point,
+	threshold int) (pubKey kyber.Point, err error) {
+
+	actor, err := c.dkg.Listen()
+	if err != nil {
+		return nil, xerrors.Errorf("failed to listen dkg: %v", err)
+	}
```

Comment:
How do you handle multiple nodes ? All of them should use Listen but only one needs to do a setup.

Reply:
That's a valid concern indeed. Could be solved by adding an argument `isLeader bool` and leave the responsibility to the user to only call Setup once with isLeader = true.
Then we should update the dkg.Actor to add a `GetPublicKey()` function, because right now the only way to get get the public key is by calling `actor.Setup()`.

Reply:
Yes with `GetPublicKey()` but the calypso interface should provide a `Listen` and a `Setup` separated.